TypeScript user first impressions - NodeJS docs don't pass type-checking

```typescript
const sql = postgres({
  host: "xdtb",
  port: 5432,
  fetch_types: false, // currently required https://github.com/xtdb/xtdb/issues/3607
  types: {
    bool: {to: 16},
    int: {
        to: 20,
        from: [23, 20], // int4, int8
        parse: parseInt
    }
  }
});
```

fails with

```
src/index.ts(78,25): error TS2769: No overload matches this call.
  Overload 2 of 2, '(url: string, options?: Options<{}> | undefined): Sql<{}>', gave the following error.
    Argument of type '{ host: string; port: number; fetch_types: boolean; types: { bool: { to: number; }; int: { to: number; from: number[]; parse: (string: string, radix?: number) => number; }; }; }' is not assignable to parameter of type 'string'.
```

```typescript
const xtdbPg = Postgres(xtdbPostgresUrl, {
  host: "xtdb",
  port: 5432,
  fetch_types: false, // currently required https://github.com/xtdb/xtdb/issues/3607
  types: {
    bool: {to: 16},
    int: {
        to: 20,
        from: [23, 20], // int4, int8
        parse: parseInt
    }
  }
});
```

fails with

```
src/index.ts(83,5): error TS2739: Type '{ to: number; }' is missing the following properties from type 'PostgresType<any>': from, serialize, parse
src/index.ts(84,5): error TS2741: Property 'serialize' is missing in type '{ to: number; from: number[]; parse: (string: string, radix?: number) => number; }' but required in type 'PostgresType<any>'.
```

Thanks @tempire for the heads-up 🙏

--- a/docs/src/content/docs/drivers/nodejs.md
+++ b/docs/src/content/docs/drivers/nodejs.md
@@ -2,35 +2,103 @@
 title: Using XTDB from JavaScript
 ---
 
-In NodeJS, you can talk to a running XTDB node using the ['postgres'](https://www.npmjs.com/package/postgres) package, taking advantage of XTDB's PostgreSQL wire-compatibility.
+In NodeJS, you can talk to a running XTDB node using the [postgres](https://www.npmjs.com/package/postgres) package, taking advantage of XTDB's PostgreSQL wire-compatibility.
+
+This is a basic configuration for XTDB interoperability:
 
 ``` javascript
 "use strict"
 
 import postgres from 'postgres';
 
+const OID = {
+  boolean: 16,
+  int64: 20,
+  int32: 23,
+  text: 25,
+  float64: 701,
+  transit: 16384,
+};
+
 const sql = postgres({
   host: "localhost",
   port: 5432,
-  fetch_types: false, // currently required https://github.com/xtdb/xtdb/issues/3607
-  types: {
-    bool: {to: 16},
-    int: {
-        to: 20,
-        from: [23, 20], // int4, int8
-        parse: parseInt
-    }
-  }
 });
 
 async function main() {
-    await sql`INSERT INTO users (_id, name) VALUES (${sql.typed.int(1)}, 'James'), (${sql.typed.int(2)}, 'Jeremy')`
+    await sql`
+      INSERT INTO users (_id, name) VALUES
+      (${sql.typed(1, OID.int32)}, ${sql.typed("James", OID.text)}),
+      (${sql.typed(2, OID.int32)}, ${sql.typed("Jeremy", OID.text)})
+    `;
 
-    console.log([...await sql`SELECT _id, name FROM users`])
-    // => [{_id: 1, name: "James"}, [{_id: 2, name: "Jeremy"}]]
+    console.log([...(await sql`SELECT _id, name FROM users`)]);
+    // => [ { _id: 2, name: 'Jeremy' }, { _id: 1, name: 'James' } ]
 }
 
 main();
+```
+
+## Specifying the data type of parameters
+
+XTDB learns your DB schema as you enter data. It doesn't provide a way for specifying and enforcing a schema in advance.
+
+Therefore, you will want to choose a PostgreSQL client that supports specifying the data type of inserted columns. This is done in PostgreSQL through the use of [OIDs](https://www.postgresql.org/docs/current/datatype-oid.html), which are just numbers that identify a DB data type.
+
+As shown in the example above, the [postgres](https://www.npmjs.com/package/postgres) package does support specifying OIDs for types. (At the time of this writing, that is not the case for the [pg](https://www.npmjs.com/package/pg) package).
+
+## Ensuring parameters and returned records are fully typed
+
+XTDB supports `transit`, which is a structured data format with typed values.
+
+This is a sample configuration of `Postgres.js` for transit support:
+
+``` javascript
+import transit from "transit-js";
+
+
+const transitReader = transit.reader("json");
+const transitWriter = transit.writer("json");
+
+const sql = postgres({
+  ...,
+
+  connection: {
+    // Record objects will be returned fully typed using the transit format:
+    fallback_output_format: "transit",
+  },
+
+  types: {
+    // Add support for the transit format:
+    transit: {
+      to: 16384,
+      from: [16384],
+      serialize: (v) => transitWriter.write(v),
+      parse: (v) => transitReader.read(v),
+    },
+
+    // By default, int64 values are handled as text.
+    // Reading int64 values as a number, ensuring no loss of precision:
+    int64: {
+      from: [20],
+      parse: (x) => {
+        const res = parseInt(x);
+        if (!Number.isSafeInteger(res))
+          throw Error(`Could not convert to integer reliably: ${x}`);
+        return res;
+      },
+    } /*as unknown as postgres.PostgresType*/, // for TypeScript
+  },
+});
+```
+
+The above configuration allows to pass record parameters as fully typed objects:
+``` javascript
+await sql`
+  INSERT INTO users (_id, name) RECORDS
+    ${sql.types.transit({ _id: 1, name: "James" })},
+    ${sql.types.transit({ _id: 2, name: "Jeremy" })}
+`;
 ```
 
 ## Examples

--- a/lang/js/test/docs.js
+++ b/lang/js/test/docs.js
@@ -1,0 +1,113 @@
+import postgres from "postgres";
+import * as uuid from "uuid";
+import transit from "transit-js";
+import assert from "assert";
+
+describe("code examples", function () {
+  const OID = {
+    boolean: 16,
+    int64: 20,
+    int32: 23,
+    text: 25,
+    float64: 701,
+    transit: 16384,
+  };
+
+  const commonPostgresOptions = () => ({
+    host: process.env.PG_HOST || "localhost",
+    port: process.env.PG_PORT || 5439,
+    database: uuid.v4().toString(),
+  });
+
+  it("example 1", async () => {
+    const sql = postgres(commonPostgresOptions());
+
+    await sql`
+      INSERT INTO users (_id, name) VALUES
+      (${sql.typed(1, OID.int32)}, ${sql.typed("James", OID.text)}),
+      (${sql.typed(2, OID.int32)}, ${sql.typed("Jeremy", OID.text)})
+    `;
+
+    console.log([...(await sql`SELECT _id, name FROM users`)]);
+    // => [ { _id: 2, name: 'Jeremy' }, { _id: 1, name: 'James' } ]
+
+    assert.deepStrictEqual(
+      [...(await sql`SELECT _id, name FROM users ORDER BY _id`)],
+      [
+        { _id: 1, name: "James" },
+        { _id: 2, name: "Jeremy" },
+      ],
+    );
+  });
+
+  it("example 2", async () => {
+    const transitReader = transit.reader("json");
+    const transitWriter = transit.writer("json");
+
+    const sql = postgres({
+      ...commonPostgresOptions(),
+
+      connection: {
+        // Record objects will be returned fully typed using the transit format:
+        fallback_output_format: "transit",
+      },
+
+      types: {
+        // Add support for the transit format:
+        transit: {
+          to: 16384,
+          from: [16384],
+          serialize: (v) => transitWriter.write(v),
+          parse: (v) => transitReader.read(v),
+        },
+
+        // By default, Postgres.js reads int64 as text.
+        // Reading as JS number, ensuring no loss of precision:
+        int64: {
+          from: [20],
+          parse: (x) => {
+            const res = parseInt(x);
+            if (!Number.isSafeInteger(res))
+              throw Error(`Could not convert to number: ${x}`);
+            return res;
+          },
+        } /*as unknown as postgres.PostgresType*/, // [add for TypeScript]
+      },
+    });
+
+    await sql`
+      INSERT INTO users (_id, name) RECORDS
+        ${sql.types.transit({ _id: 1, name: "James" })},
+        ${sql.types.transit({ _id: 2, name: "Jeremy" })}
+    `;
+
+    console.log([...(await sql`SELECT _id, name FROM users`)]);
+
+    assert.deepStrictEqual(
+      [...(await sql`SELECT _id, name FROM users ORDER BY _id`)],
+      [
+        { _id: 1, name: "James" },
+        { _id: 2, name: "Jeremy" },
+      ],
+    );
+  });
+});
+
+// Further code excerpts that could be added to docs in the future:
+//
+//     const transitReader = transit.reader("json", {
+//       mapBuilder: {
+//         init: () => ({}),
+//         finalize: (m) => m,
+//         add: (m, k, v) => ({ ...m, [k]: v }),
+//       },
+//       handlers: {
+//         "time/zoned-date-time": (rep) => {
+//           // We receive "2024-11-06T11:43:20.123Z[UTC]"
+//           // See https://docs.oracle.com/javase/8/docs/api/java/time/format/DateTimeFormatter.html#ISO_ZONED_DATE_TIME
+//           const withoutTz = rep.replace(/\[[^\]]*]$/, "");
+//           return new Date(withoutTz);
+//         },
+//         "time/duration": (rep) => rep,
+//       },
+//     });

